### PR TITLE
Fix the offset without multiplier or multiplier without offset bug

### LINKED ISSUE #497: Issue with offset/multiplier if no offset is specified
Compare the outputs of these three models (20000 posterior draws, 4 chains each).

Manual non-centering:

```
parameters {
  real<lower = 0.0> sigma;
  real a_raw;
}

transformed parameters {
  real a = sigma * a_raw;
}

model {
  sigma ~ normal(0, 1);
  a_raw ~ normal(0, 1);
}
```

```
# A tibble: 4 x 10
  variable     mean    median    sd   mad      q5    q95  rhat ess_bulk ess_tail
  <chr>       <dbl>     <dbl> <dbl> <dbl>   <dbl>  <dbl> <dbl>    <dbl>    <dbl>
1 lp__     -1.64    -1.30     1.11  0.823 -3.86   -0.559  1.00   22611.   24752.
2 sigma     0.795    0.673    0.602 0.595  0.0603  1.96   1.00   24692.   18954.
3 a_raw    -0.00312 -0.00212  0.996 0.988 -1.64    1.64   1.00   34992.   38876.
4 a        -0.00538 -0.000228 0.991 0.533 -1.59    1.57   1.00   33141.   41470.
```

Multiplier, no offset:

```
parameters {
  real<lower = 0.0> sigma;
  real<multiplier = sigma> a;
}

transformed parameters {
  real a_raw = a / sigma;
}

model {
  sigma ~ normal(0, 1);
  a ~ normal(0, sigma);
}
```

```
# A tibble: 4 x 10
  variable      mean     median    sd   mad     q5     q95  rhat ess_bulk ess_tail
  <chr>        <dbl>      <dbl> <dbl> <dbl>  <dbl>   <dbl> <dbl>    <dbl>    <dbl>
1 lp__     -1.01     -0.704     1.02  0.726 -3.08  -0.0538  1.00    2767.    1868.
2 sigma     0.829     0.705     0.595 0.584  0.111  1.97    1.01    1509.     507.
3 a         0.000879 -0.0000682 1.03  0.573 -1.65   1.62    1.00   18306.   14657.
4 a_raw    -0.00257  -0.000104  0.993 0.973 -1.64   1.63    1.00   15091.   10312.
```

Offset and multiplier:

```
parameters {
  real<lower = 0.0> sigma;
  real<offset = 0.0, multiplier = sigma> a;
}

transformed parameters {
  real a_raw = a / sigma;
}

model {
  sigma ~ normal(0, 1);
  a ~ normal(0, sigma);
}
```

```
# A tibble: 4 x 10
  variable     mean    median    sd   mad      q5    q95  rhat ess_bulk ess_tail
  <chr>       <dbl>     <dbl> <dbl> <dbl>   <dbl>  <dbl> <dbl>    <dbl>    <dbl>
1 lp__     -1.63    -1.30     1.10  0.816 -3.82   -0.561  1.00   23740.   26454.
2 sigma     0.797    0.675    0.601 0.587  0.0651  1.95   1.00   27799.   20308.
3 a         0.00632  0.000170 0.985 0.547 -1.58    1.61   1.00   33588.   39913.
4 a_raw     0.00137  0.000993 0.999 1.01  -1.64    1.65   1.00   35724.   39686.
```

Here is code to reinstall cmdstanr from develop:
```
library(cmdstanr)
install_cmdstan(cores = 8, overwrite = TRUE, repo_clone = TRUE)
```

Here is code to run the models (assuming you have them saved as `test1.stan`, `test2.stan`, and `test3.stan`):

```
m1 = cmdstan_model("test1.stan")
m2 = cmdstan_model("test2.stan")
m3 = cmdstan_model("test3.stan")

f1 = m1$sample(num_samples = 20000)
f2 = m2$sample(num_samples = 20000)
f3 = m3$sample(num_samples = 20000)

f1$summary()
f2$summary()
f3$summary()
```

### PR DESCRIPTION
**Pull request by rok-cesnovar**
Fixes #497

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Does this also let them occur in either order?  If not, would that be hard to add?  I don't want to hold up the fix, but it should also be order independent.

Could you also check that `lower` and `upper` don't repeat the same bug?

Thanks!

**Comment by rok-cesnovar** [comment]
> Does this also let them occur in either order?

I dont think so. If I do that, i get: 

```
Syntax error in 'examples/bugs/off_mult1.stan', line 3, column 20 to column 25, parsing error:
   -------------------------------------------------
     1:  parameters {
     2:    real<lower = 0.0> sigma;
     3:    real<multiplier = sigma, offset = 0.0> a[5];
                                  ^
     4:  }
     5:  
   -------------------------------------------------

Expected ">" after "multiplier =" expression.
```
I suspect the fix is easy, but I have so far clawed my way from basic understanding of generating c++ to partially understanding the MIR representation. I am not friends with the parser yet :/ Maybe someone else can chime in.

> Could you also check that `lower` and `upper` don't repeat the same bug?

Just lower produces -> sigma = stan::math::lb_constrain(sigma, 0.0);
Just upper produces -> sigma = stan::math::ub_constrain(sigma, 1.0);
Both produces -> sigma = stan::math::lub_constrain(sigma, 0.0, 1.0);

So that is fine. But also order dependent.

**Comment by bob-carpenter** [comment]
Thanks for checking.  I created issue #511 for the order dependence.

**Comment by rok-cesnovar** [comment]
It would be great to get this fix in for the release if anyone has time to have a look.

File: src/frontend/Ast_to_Mir.ml [hunk 1]
```diff
@@ -257,8 +257,8 @@ let constraint_forl = function
 
 let extract_transform_args = function
   | Program.Lower a | Upper a -> [a]
-  | Offset a -> [{a with Expr.Fixed.pattern= Lit (Int, "0")}; a]
-  | Multiplier a -> [a; {a with pattern= Lit (Int, "1")}]
+  | Offset a -> [a; {a with Expr.Fixed.pattern= Lit (Int, "1")}]
+  | Multiplier a -> [{a with pattern= Lit (Real, "0.0")}; a]
```
Comment (nhuurre) [inline src/frontend/Ast_to_Mir.ml:261]: Why is one `Int` and the other `Real`? It's not like the backend cares.

Comment (rok-cesnovar) [reply src/frontend/Ast_to_Mir.ml:261]: Good call. Changing to Int 0 and merging. Thanks!

**Pull request merged**